In [1]:
import pandas as pd
pd.set_option("display.max_columns", 101)

# Clean and prepare dataset

We create a dataframe named df_pivot that contains job features (skills and knowledge) as columns and job titles as rows. For the purpose of this project, jobs are filtered depending on their job zones. 
All the data can be found here:
https://www.onetcenter.org/database.html

In [2]:
# Load datasets into Pandas DataFrame
skills = pd.read_excel('../databases/ONET/Skills.xlsx')

In [3]:
knowledge = pd.read_excel('../databases/ONET/Knowledge.xlsx')

In [4]:
# Merge skills and knowledge files
df = skills.append(knowledge)

In [5]:
# Filter per job zones 3, 4 and 5
job_zones = pd.read_excel('../databases/ONET/Job Zones.xlsx')
jobzone345 = job_zones[job_zones['Job Zone'].isin([3,4,5])]['O*NET-SOC Code'].tolist()
df = df[df['O*NET-SOC Code'].isin(jobzone345)]

In [6]:
# Take out unnecessary columns
cols_to_keep = [
    'O*NET-SOC Code', 
    'Title', 
    'Element Name', 
    'Scale ID',
    'Data Value'
]

df = df[cols_to_keep].copy()

In [7]:
# Merge columns 'Element Name' and 'Scale ID' into 'element_scale_id'
df['element_scale_id'] = df['Element Name'] + '_' + df['Scale ID'] 

In [8]:
# Pivot table
df_pivot = df.pivot_table(index=['Title'], columns='element_scale_id', values='Data Value')

In [9]:
df_pivot.head()

element_scale_id                        Active Learning_IM  \
Title                                                        
Accountants                                           3.25   
Actuaries                                             3.38   
Acupuncturists                                        3.12   
Acute Care Nurses                                     3.75   
Adapted Physical Education Specialists                3.50   

element_scale_id                        Active Learning_LV  \
Title                                                        
Accountants                                           3.38   
Actuaries                                             4.12   
Acupuncturists                                        3.50   
Acute Care Nurses                                     3.88   
Adapted Physical Education Specialists                3.62   

element_scale_id                        Active Listening_IM  \
Title                                                         
Accountants                                            3.88   
Actuaries                                              4.00   
Acupuncturists                                         3.75   
Acute Care Nurses                                      4.00   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Active Listening_LV  \
Title                                                         
Accountants                                            3.62   
Actuaries                                              4.12   
Acupuncturists                                         3.62   
Acute Care Nurses                                      3.88   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Administration and Management_IM  \
Title                                                                      
Accountants                                                         3.43   
Actuaries                                                           3.31   
Acupuncturists                                                      2.67   
Acute Care Nurses                                                   2.89   
Adapted Physical Education Specialists                              2.44   

element_scale_id                        Administration and Management_LV  \
Title                                                                      
Accountants                                                         4.00   
Actuaries                                                           4.72   
Acupuncturists                                                      2.57   
Acute Care Nurses                                                   2.96   
Adapted Physical Education Specialists                              2.56   

element_scale_id                        Biology_IM  Biology_LV  \
Title                                                            
Accountants                                   1.02        0.03   
Actuaries                                     1.16        0.47   
Acupuncturists                                3.45        4.05   
Acute Care Nurses                             3.32        3.50   
Adapted Physical Education Specialists        2.65        2.81   

element_scale_id                        Building and Construction_IM  \
Title                                                                  
Accountants                                                     1.21   
Actuaries                                                       1.25   
Acupuncturists                                                  1.60   
Acute Care Nurses                                               1.18   
Adapted Physical Education Specialists                          1.58   

element_scale_id                        Building and Construction_LV  \
Title                                                                  
Accountants                                                     0.40

In [10]:
# Display size of the matrix
df_pivot.shape

(640, 134)

Let's factorize this matrix to (640, 2)

# Create the embedding with UMAP

We create an embedding with UMAP and project it in 2D.

In [11]:
import umap

In [12]:
# Create embedding
embedding = umap.UMAP(n_neighbors=15, n_components=2, random_state=42,
                      min_dist=0.3,
                      metric='correlation').fit_transform(df_pivot.values)

In [13]:
# Display shape of embedding
embedding.shape

(640, 2)

In [14]:
embedding

array([[ 1.5840942 , -4.695757  ],
       [ 1.1996833 , -4.093121  ],
       [ 3.6107712 ,  0.3655267 ],
       ...,
       [-1.7982806 , -2.4191442 ],
       [-6.2889442 , -4.4923296 ],
       [-0.7414811 , -0.30035707]], dtype=float32)

We now have an array of floats that correspond to the coordinates of each point of the dataset in the target space. Let's visualize it.

# Visualization with plotly

In [15]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)
import plotly.graph_objs as go

In [16]:
data = [
    go.Scatter(
        x=embedding[:,0],
        y=embedding[:,1],
        mode='markers',
        text=df_pivot.index.tolist(),
        opacity=0.5
    )
]
layout = go.Layout(
    title='Jobs similarity map (based on skills and knowledge) with UMAP',
    hovermode='closest'
)
fig = go.Figure(data=data, layout=layout)
plot_url = iplot(fig, filename='hover-chart-basic')

Thanks!

# Json of coordinates


We create a DataFrame with each marker's id (O*NET-SOC Code), title, coordinates (based on the UMAP embedding) and description. We export this DataFrame into a Json that we'll use to plot markers on a map.

In [42]:
# Create a series with each occupation's O*NET-SOC Code as the index and the corresponding Title as the value
id_title = skills.sort_values('Title').drop_duplicates('Title').set_index('O*NET-SOC Code').Title

In [43]:
id_title.head()

O*NET-SOC Code
13-2011.01          Accountants
27-2011.00               Actors
15-2011.00            Actuaries
29-1199.01       Acupuncturists
29-1141.01    Acute Care Nurses
Name: Title, dtype: object

In [44]:
# Convert this series into a DataFrame
export = pd.DataFrame(id_title)

In [23]:
# To avoid typos, let's rename the 'Title' column:'title'
export = export.rename(columns={'Title': 'title'})

In [24]:
export.head()

title
O*NET-SOC Code                   
13-2011.01            Accountants
27-2011.00                 Actors
15-2011.00              Actuaries
29-1199.01         Acupuncturists
29-1141.01      Acute Care Nurses

In [25]:
# Create a dictionary of all titles and their corresponding array of coordinates based on the UMAP embedding
title_embedding = dict(zip(df_pivot.index.tolist(), embedding))

In [26]:
title_embedding

{'Accountants': array([ 1.5840942, -4.695757 ], dtype=float32),
 'Actuaries': array([ 1.1996833, -4.093121 ], dtype=float32),
 'Acupuncturists': array([3.6107712, 0.3655267], dtype=float32),
 'Acute Care Nurses': array([2.9483    , 0.21456927], dtype=float32),
 'Adapted Physical Education Specialists': array([ 4.0689173, -1.5566826], dtype=float32),
 'Administrative Law Judges, Adjudicators, and Hearing Officers': array([ 2.4501   , -3.9124715], dtype=float32),
 'Administrative Services Managers': array([ 0.6157187, -3.492248 ], dtype=float32),
 'Adult Basic and Secondary Education and Literacy Teachers and Instructors': array([ 4.3629856, -2.4391916], dtype=float32),
 'Advanced Practice Psychiatric Nurses': array([3.9976122 , 0.11710694], dtype=float32),
 'Advertising Sales Agents': array([ 0.32229546, -5.011517  ], dtype=float32),
 'Advertising and Promotions Managers': array([-0.6765995, -4.857364 ], dtype=float32),
 'Aerospace Engineering and Operations Technicians': array([-5.1346

In [28]:
# Map 'export' with 'title_embedding' based on the title column & add column with coordinates
export['coords'] = export.title.map(title_embedding)

In [29]:
export.head()

title                  coords
O*NET-SOC Code                                           
13-2011.01            Accountants  [1.5840942, -4.695757]
27-2011.00                 Actors                     NaN
15-2011.00              Actuaries  [1.1996833, -4.093121]
29-1199.01         Acupuncturists  [3.6107712, 0.3655267]
29-1141.01      Acute Care Nurses    [2.9483, 0.21456927]

In [45]:
# Load 'Occupation Data.xlsx' into Pandas DataFrame
occupation_data = pd.read_excel('../databases/ONET/Occupation Data.xlsx')

In [47]:
occupation_data.head()

O*NET-SOC Code                                Title  \
0     11-1011.00                     Chief Executives   
1     11-1011.03        Chief Sustainability Officers   
2     11-1021.00      General and Operations Managers   
3     11-1031.00                          Legislators   
4     11-2011.00  Advertising and Promotions Managers   

                                         Description  
0  Determine and formulate policies and provide o...  
1  Communicate and coordinate with management, sh...  
2  Plan, direct, or coordinate the operations of ...  
3  Develop, introduce or enact laws and statutes ...  
4  Plan, direct, or coordinate advertising polici...

In [31]:
# Set O*NET-SOC Code as the index and Description as the value of a series named 'description'
description = occupation_data.set_index('O*NET-SOC Code').Description

In [32]:
description

O*NET-SOC Code
11-1011.00    Determine and formulate policies and provide o...
11-1011.03    Communicate and coordinate with management, sh...
11-1021.00    Plan, direct, or coordinate the operations of ...
11-1031.00    Develop, introduce or enact laws and statutes ...
11-2011.00    Plan, direct, or coordinate advertising polici...
11-2011.01    Create and implement methods to market green p...
11-2021.00    Plan, direct, or coordinate marketing policies...
11-2022.00    Plan, direct, or coordinate the actual distrib...
11-2031.00    Plan, direct, or coordinate activities designe...
11-3011.00    Plan, direct, or coordinate one or more admini...
11-3021.00    Plan, direct, or coordinate activities in such...
11-3031.00    Plan, direct, or coordinate accounting, invest...
11-3031.01    Direct financial activities, such as planning,...
11-3031.02    Direct and coordinate financial activities of ...
11-3051.00    Plan, direct, or coordinate the work activitie...
11-3051.01    Plan, direc

In [48]:
# Map 'description' and 'export' based on index
export['description'] = export.index.map(lambda x: description[x])

In [49]:
export

Title  \
O*NET-SOC Code                                                      
13-2011.01                                            Accountants   
27-2011.00                                                 Actors   
15-2011.00                                              Actuaries   
29-1199.01                                         Acupuncturists   
29-1141.01                                      Acute Care Nurses   
25-2059.01                 Adapted Physical Education Specialists   
51-9191.00         Adhesive Bonding Machine Operators and Tenders   
23-1021.00      Administrative Law Judges, Adjudicators, and H...   
11-3011.00                       Administrative Services Managers   
25-3011.00      Adult Basic and Secondary Education and Litera...   
29-1141.02                   Advanced Practice Psychiatric Nurses   
41-3011.00                               Advertising Sales Agents   
11-2011.00                    Advertising and Promotions Managers   
17-3021.00       Aerospace Engineering and Operations Technicians   
17-2011.00                                    Aerospace Engineers   
13-1011.00      Agents and Business Managers of Artists, Perfo...   
17-2021.00                                 Agricultural Engineers   
45-2091.00                       Agricultural Equipment Operators   
45-2011.00                                Agricultural Inspectors   
25-1041.00          Agricultural Sciences Teachers, Postsecondary   
19-4011.01                               Agricultural Technicians   
53-2021.00                                Air Traffic Controllers   
53-1011.00                    Aircraft Cargo Handling Supervisors   
49-3011.00             Aircraft Mechanics and Service Technicians   
51-2011.00      Aircraft Structure, Surfaces, Rigging, and Sys...   
53-2022.00                        Airfield Operations Specialists   
53-2011.00         Airline Pilots, Copilots, and Flight Engineers   
29-1069.01                           Allergists and Immunologists   
53-3011.00      Ambulance Drivers and Attendants, Except Emerg...   
39-3091.00                    Amusement and Recreation Attendants   
...                                                           ...   
51-6093.00                                           Upholsterers   
19-3051.00                            Urban and Regional Planners   
29-1069.12                                             Urologists   
39-3031.00            Ushers, Lobby Attendants, and Ticket Takers   
17-2199.02                                   Validation Engineers   
29-1131.00                                          Veterinarians   
31-9096.00      Veterinary Assistants and Laboratory Animal Ca...   
29-2056.00               Veterinary Technologists and Technicians   
15-1199.11                                   Video Game Designers   
25-1194.00           Vocational Education Teachers, Postsecondary   
35-3031.00                                 Waiters and Waitresses   
49-9064.00                                        Watch Repairers   
11-9121.02                             Water Resource Specialists   
51-8031.00      Water and Wastewater Treatment Plant and Syste...   
17-2081.01                             Water/Wastewater Engineers   
47-4099.03              Weatherization Installers and Technicians   
15-1199.03                                     Web Administrators   
15-1134.00                                         Web Developers   
43-5111.00      Weighers, Measurers, Checkers, and Samplers, R...   
51-4121.06                   Welders, Cutters, and Welder Fitters   
51-4122.00      Welding, Soldering, and Brazing Machine Setter...   
53-7073.00                                       Wellhead Pumpers   
13-1022.00      Wholesale and Retail Buyers, Except Farm Products   
17-2199.10                                  Wind Energy Engineers   
11-9199.09                        Wind Energy Operations Managers   
11-9199.10                           Wind Energy Project M

In [39]:
type(export)

pandas.core.frame.DataFrame

In [45]:
# Convert 'export' into json
export.to_json('map_data.json', orient='index')